src/imageio/imageio_png.c

```diff
@@ -199,18 +199,31 @@ dt_imageio_retval_t dt_imageio_open_png(dt_image_t *img, const char *filename, d
     return DT_IMAGEIO_LOAD_FAILED;
   }
 
-  for(size_t j = 0; j < height; j++)
-  {
-    if(bpp < 16)
-      for(size_t i = 0; i < width; i++)
-        for(int k = 0; k < 3; k++)
-          mipbuf[4 * (j * width + i) + k] = buf[3 * (j * width + i) + k] * (1.0f / 255.0f);
-    else
-      for(size_t i = 0; i < width; i++)
-        for(int k = 0; k < 3; k++)
-          mipbuf[4 * (j * width + i) + k] = (256.0f * buf[2 * (3 * (j * width + i) + k)]
-                                             + buf[2 * (3 * (j * width + i) + k) + 1]) * (1.0f / 65535.0f);
-  }
+  if(bpp < 16)
+#ifdef _OPENMP
+#pragma omp parallel for default(none) \
+  dt_omp_firstprivate(width, height, buf) \
+  shared(mipbuf) \
+  collapse(3)
+#endif
+    for(size_t line = 0; line < height; line++)
+      for(size_t x = 0; x < width; x++)
+        for(int channel = 0; channel < 3; channel++)
+          mipbuf[4 * (line * width + x) + channel] =
+             buf[3 * (line * width + x) + channel] * (1.0f / 255.0f);
```

Comment:
It's worth benchmarking
```
for(size_t pixel = 0; pixel <width * heigh; pixel++)
   for(size_t channel = 0; channel < 3; channel++)
      mipbuf[4 * pixel + channel] = buf[3 * pixel + channel] * (1.0f / 255.0f);
```
without a `collapse` directive.  The collapse directive generates one big loop and uses divisions to slice out the original loop variables.

Reply:
@ralfbrown Thank you for the idea. I got interesting results and, indeed, "complete collapse" of all loops was not the best option.

First, comparing the results (all values are the average of the same number of launches):

- My initial version, `collapse (3)` -- `0.065` sec
- Simply enabling OpenMP without loop collapse (essentially parallelization of external cycle only, the same as if `collapse(1)` was specified) -- `0.050` sec
- `collapse(2)` - `0.045` sec

So collapse(2) is optimal here. Also, I want to try to flatten the pixel loop and see if I get an additional speed increase.

After thinking about it, I'm starting to understand that less speedup for the "lighter" 8-bit data conversion loop should be expected. Having a loop where each iteration is fast should result in less speedup, even if there are many iterations, because the parallelization overhead becomes noticeable against the fast body of the loop. This explains why when processing 16-bit data, the speedup is greater and actually close to the number of CPU cores, because there the body of the loop contains more calculations.

In the process of benchmarking, I noticed a strange situation for which I have no explanation. About it in the main flow of comments of PR, for more visibility.

Reply:
With speculative execution, auto-vectorization, caching, etc. it's impossible to definitively predict which variation will actually be fastest unless you benchmark all of them....

I would expect flattening the pixel looping (outer two loops) to give a *small* additional speed boost since some overhead can be eliminated compared to the collapse(2) version.  But as I just said, it will be necessary to benchmark for confirmation.